File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
import sys
from collections import deque

def bfs(start_x, start_y, end_x, end_y, l):
    dx = [2, 2, -2, -2, 1, 1, -1, -1]
    dy = [1, -1, 1, -1, 2, -2, 2, -2]
    n, m = l, l

    visited = {(start_x, start_y)}  # 방문 집합 생성
    queue = deque([(start_x, start_y, 0)])  # 큐에 (x, y, 거리) 넣기

    while queue:
        x, y, dist = queue.popleft()

        if x == end_x and y == end_y:  # 탐색 종료
            return dist

        for i in range(8):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < n and 0 <= ny < m: # 배열 범위 내에 있는지
                if (nx, ny) not in visited:
                    visited.add((nx, ny))
                    queue.append((nx, ny, dist + 1))  # (다음좌표x, 다음좌표y, 현재거리 + 1)

    return -1  # 큐가 비었는데 도달 못했을 경우

T = int(input())
for _ in range(T):
    l = int(input())
    start_x, start_y = map(int, sys.stdin.readline().split())
    end_x, end_y = map(int, sys.stdin.readline().split())
    print(bfs(start_x, start_y, end_x, end_y, l))

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.cc
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

/*

*/

int dy[] = {2, 2, -2, -2, 1, 1, -1, -1};
int dx[] = {1, -1, 1, -1, 2, -2, 2, -2};

int l;
int start_x, start_y, end_x, end_y;


int bfs(){
    // 초기화 반드시 해줘야함
    int dist[300][300] = {0};
    bool visited[300][300] = {false};

    queue<pair<int,int>> q;
    q.push({start_x,start_y});
    visited[start_x][start_y]=true;
    
    while (!q.empty()){
        int x = q.front().first;
        int y = q.front().second;
        int cur_dist = dist[x][y];
        q.pop();
        
        if (x==end_x && y==end_y){
            return cur_dist;
        }

        for (int i=0; i<8; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if ( nx<0 || nx>=l || ny<0 || ny>=l ) continue;
            if (visited[nx][ny]) continue;
            visited[nx][ny] = true;
            dist[nx][ny] = cur_dist + 1;
            q.push({nx, ny});
        }
    }
    return -1;  // 도달 못했을 경우
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(0);

    int T;
    cin >> T;
    while(T--){

        cin >> l;
        cin >> start_x >> start_y;
        cin >> end_x >> end_y;

        cout << bfs() << '\n';
    }
    return 0;
}